File: tf_adapter/python/npu_bridge/estimator/npu_ops.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (c) Huawei Technologies Co., Ltd. 2019-2021. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

"""Ops for collective operations implemented using hccl."""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numbers

from tensorflow.python.ops.nn_ops import _get_noise_shape
from tensorflow.python.framework import ops
from tensorflow.python.framework import random_seed
from tensorflow.python.eager import context

from npu_bridge.helper import helper
from npu_bridge.estimator.npu.npu_common import NPUBasics

gen_npu_ops = helper.get_gen_ops()

DEFAULT_GRAPH_SEED = 87654321
_MAXINT32 = 2 ** 31 - 1


def NPUInit(name=None):
    """Initiate NPU"""
    if context.executing_eagerly():
        raise RuntimeError("tf.NPUInit() is not compatible with "
                           "eager execution.")

    return gen_npu_ops.npu_init(name=name)


def NPUShutdown(name=None):
    """Shutdown a distributed NPU system for use with TensorFlow.

    Args:
        name: Name of ops.

    Returns:
        The npu init ops which will open the NPU system using `Session.run`.
    """
    if context.executing_eagerly():
        raise RuntimeError("tf.NPUShutdown() is not compatible with "
                           "eager execution.")

    return gen_npu_ops.npu_shutdown(name=name)


def initialize_system(name=None):
    """Initializes a distributed NPU system for use with TensorFlow.

    Args:
        name: Name of ops.

    Returns:
        The npu init ops which will open the NPU system using `Session.run`.
    """
    return NPUInit(name)


def shutdown_system(name=None):
    """Shuts down a running NPU system."""

    return NPUShutdown(name)


def LARS(inputs_w, inputs_g, weight_decay, hyperpara=0.001, epsilon=0.00001, name=None):
    """NPU implemented LARS"""
    if context.executing_eagerly():
        raise RuntimeError("tf.LARS() is not compatible with "
                           "eager execution.")

    return gen_npu_ops.lars(inputs_w=inputs_w, inputs_g=inputs_g, weight_decay=weight_decay, hyperpara=hyperpara,
                            epsilon=epsilon, name=name)


def LARSV2(input_weight,
           input_grad,
           weight_decay,
           learning_rate,
           hyperpara=0.001,
           epsilon=0.00001,
           use_clip=False,
           name=None):
    """NPU implemented LARSV2"""
    if context.executing_eagerly():
        raise RuntimeError("tf.LARSV2() is not compatible with "
                           "eager execution.")

    return gen_npu_ops.lars_v2(input_weight=input_weight,
                               input_grad=input_grad,
                               weight_decay=weight_decay,
                               learning_rate=learning_rate,
                               hyperpara=hyperpara,
                               epsilon=epsilon,
                               use_clip=use_clip,
                               name=name)


def outfeed_dequeue_op(channel_name, output_types, output_shapes, name=None):
    """Operator for outfeed dequeue"""
    return gen_npu_ops.outfeed_dequeue_op(channel_name=channel_name, output_types=output_types,
                                          output_shapes=output_shapes, name=name)


def outfeed_enqueue_op(channel_name, inputs, name=None):
    """Operator for outfeed enqueue"""
    return gen_npu_ops.outfeed_enqueue_op(inputs=inputs, channel_name=channel_name, name=name)


def stop_outfeed_dequeue_op(channel_name, name=None):
    """Operator for stoping outfeed dequeue"""
    return gen_npu_ops.stop_outfeed_dequeue_op(channel_name, name)


def _truncate_seed(seed):
    return seed % _MAXINT32  # Truncate to fit into 32-bit integer


def dropout(x, keep_prob, noise_shape=None, seed=None, name=None):
    """The gradient for `gelu`.

    Args:
        x: A tensor with type is float.
        keep_prob: A tensor, float, rate of every element reserved.
        noise_shape: A 1-D tensor, with type int32, shape of keep/drop what random
            generated.
        seed: Random seed.
        name: Layer name.

    Returns:
        A tensor.
    """
    if context.executing_eagerly():
        raise RuntimeError("npu_ops.dropout() is not compatible with "
                           "eager execution.")
    x = ops.convert_to_tensor(x, name="x")
    if not x.dtype.is_floating:
        raise ValueError("x must be a floating point tensor."
                         " Got a %s tensor instead." % x.dtype)
    if isinstance(keep_prob, numbers.Real) and not 0 < keep_prob <= 1.0:
        raise ValueError("keep_prob must be a float value or a scalar tensor in the "
                         "range (0, 1], got %g" % keep_prob)
    if isinstance(keep_prob, float) and keep_prob == 1.0:
        return x
    seed, seed2 = random_seed.get_seed(seed)
    noise_shape = _get_noise_shape(x, noise_shape)
    gen_out = gen_npu_ops.drop_out_gen_mask(noise_shape, keep_prob, seed, seed2, name)
    result = gen_npu_ops.drop_out_do_mask(x, gen_out, keep_prob, name)
    return result


@ops.RegisterGradient("DropOutDoMask")
def _DropOutDoMaskGrad(op, grad):
    result = gen_npu_ops.drop_out_do_mask(grad, op.inputs[1], op.inputs[2])
    return [result, None, None]


def basic_lstm_cell(x, h, c, w, b, keep_prob, forget_bias, state_is_tuple,
                    activation, name=None):
    """NPU implemented lstm cell"""
    if context.executing_eagerly():
        raise RuntimeError("tf.basic_lstm_cell() is not compatible with "
                           "eager execution.")
    x = ops.convert_to_tensor(x, name="x")
    h = ops.convert_to_tensor(h, name="h")
    c = ops.convert_to_tensor(c, name="c")
    w = ops.convert_to_tensor(w, name="w")
    b = ops.convert_to_tensor(b, name="b")
    result = gen_npu_ops.basic_lstm_cell(x, h, c, w, b, keep_prob, forget_bias, state_is_tuple,
                                         activation, name)
    return result


@ops.RegisterGradient("BasicLSTMCell")
def basic_lstm_cell_grad(op, dct, dht, dit, djt, dft, dot, dtanhct):
    """NPU implemented gradient for lstm cell"""
    dgate, dct_1 = gen_npu_ops.basic_lstm_cell_c_state_grad(op.inputs[2], dht, dct, op.outputs[2], op.outputs[3],
                                                            op.outputs[4], op.outputs[5], op.outputs[6],
                                                            forget_bias=op.get_attr("forget_bias"),
                                                            activation=op.get_attr("activation"))
    dw, db = gen_npu_ops.basic_lstm_cell_weight_grad(op.inputs[0], op.inputs[1], dgate)
    dxt, dht = gen_npu_ops.basic_lstm_cell_input_grad(dgate, op.inputs[3], keep_prob=op.get_attr("keep_prob"))

    return [dxt, dht, dct_1, dw, db]


def adam_apply_one_assign(input0, input1, input2, input3, input4,
                          mul0_x, mul1_x, mul2_x, mul3_x, add2_y, name=None):
    """NPU implemented adam_apply_one_assign"""
    if context.executing_eagerly():
        raise RuntimeError("tf.adam_apply_one_assign() is not compatible with "
                           "eager execution.")
    result = gen_npu_ops.adam_apply_one_assign(input0, input1, input2, input3, input4,
                                               mul0_x, mul1_x, mul2_x, mul3_x, add2_y, name)
    return result


def adam_apply_one_with_decay_assign(input0, input1, input2, input3, input4,
                                     mul0_x, mul1_x, mul2_x, mul3_x, mul4_x, add2_y, name=None):
    """NPU implemented adam_apply_one_with_decay_assign"""
    if context.executing_eagerly():
        raise RuntimeError("tf.adam_apply_one_with_decay_assign() is not compatible with "
                           "eager execution.")
    result = gen_npu_ops.adam_apply_one_with_decay_assign(input0, input1, input2, input3, input4,
                                                          mul0_x, mul1_x, mul2_x, mul3_x, mul4_x, add2_y, name)
    return result


@ops.RegisterGradient("DynamicGruV2")
def dynamic_gru_v2_grad(op, dy, doutput_h, dupdate, dreset, dnew, dhidden_new):
    """NPU implemented dynamic_gru_v2"""
    (x, weight_input, weight_hidden, bias_input, bias_hidden, seq_length, init_h) = op.inputs
    (y, output_h, update, reset, new, hidden_new) = op.outputs
    (dw_input, dw_hidden, db_input, db_hidden, dx, dh_prev) = gen_npu_ops.dynamic_gru_v2_grad(x, weight_input,
                                                                                              weight_hidden, y, init_h,
                                                                                              output_h, dy, doutput_h,
                                                                                              update, reset, new,
                                                                                              hidden_new,
                                                                                              direction=op.get_attr(
                                                                                                  "direction"),
                                                                                              cell_depth=op.get_attr(
                                                                                                  "cell_depth"),
                                                                                              keep_prob=op.get_attr(
                                                                                                  "keep_prob"),
                                                                                              cell_clip=op.get_attr(
                                                                                                  "cell_clip"),
                                                                                              num_proj=op.get_attr(
                                                                                                  "num_proj"),
                                                                                              time_major=op.get_attr(
                                                                                                  "time_major"),
                                                                                              gate_order=op.get_attr(
                                                                                                  "gate_order"),
                                                                                              reset_after=op.get_attr(
                                                                                                  "reset_after"))

    return (dx, dw_input, dw_hidden, db_input, db_hidden, seq_length, dh_prev)


@ops.RegisterGradient("DynamicRnn")
def dynamic_rnn_grad(op, dy, dh, dc, di, dj, df, do, dtanhc):
    """NPU implemented dynamic_rnn_grad"""
    (x, w, b, seq_length, init_h, init_c) = op.inputs
    (y, output_h, output_c, i, j, f, o, tanhc) = op.outputs
    (dw, db, dx, dh_prev, dc_prev) = gen_npu_ops.dynamic_rnn_grad(x, w, b, y, init_h[-1], init_c[-1], output_h,
                                                                  output_c, dy, dh[-1], dc[-1], i, j, f, o, tanhc,
                                                                  cell_type=op.get_attr("cell_type"),
                                                                  direction=op.get_attr("direction"),
                                                                  cell_depth=op.get_attr("cell_depth"),
                                                                  use_peephole=op.get_attr("use_peephole"),
                                                                  keep_prob=op.get_attr("keep_prob"),
                                                                  cell_clip=op.get_attr("cell_clip"),
                                                                  num_proj=op.get_attr("num_proj"),
                                                                  time_major=op.get_attr("time_major"),
                                                                  forget_bias=op.get_attr("forget_bias"))

    return (dx, dw, db, seq_length, dh_prev, dc_prev)


@ops.RegisterGradient("DynamicRnnV2")
def dynamic_rnn_v2_grad(op, dy, dh, dc, di, dj, df, do, dtanhc):
    """NPU implemented dynamic_rnn_v2_grad"""
    (x, w, b, init_h, init_c) = op.inputs
    (y, output_h, output_c, i, j, f, o, tanhc) = op.outputs
    (dw, db, dx, dh_prev, dc_prev) = gen_npu_ops.dynamic_rnn_grad(x, w, b, y, init_h[-1], init_c[-1], output_h,
                                                                  output_c, dy, dh[-1], dc[-1], i, j, f, o, tanhc,
                                                                  cell_type=op.get_attr("cell_type"),
                                                                  direction=op.get_attr("direction"),
                                                                  cell_depth=op.get_attr("cell_depth"),
                                                                  use_peephole=op.get_attr("use_peephole"),
                                                                  keep_prob=op.get_attr("keep_prob"),
                                                                  cell_clip=op.get_attr("cell_clip"),
                                                                  num_proj=op.get_attr("num_proj"),
                                                                  time_major=op.get_attr("time_major"),
                                                                  forget_bias=op.get_attr("forget_bias"))

    return (dx, dw, db, dh_prev, dc_prev)


def scatter_elements(data, indices, updates, axis=0, name=None):
    """Scatter data based on indices"""
    data = ops.convert_to_tensor(data, name="data")
    indices = ops.convert_to_tensor(indices, name="indices")
    updates = ops.convert_to_tensor(updates, name="updates")
    y = gen_npu_ops.scatter_elements(data, indices, updates, axis, name)
    return y


def k_means_centroids(x, y, sum_square_y, sum_square_x, use_actual_distance=False, name=None):
    """k_means_centroids.

    Args:
        x: A tensor with type is float.
        y: A tensor with type is float.
        sum_square_y: A tensor with type is float.
        sum_square_x: A tensor with type is float or None.
        use_actual_distance: Whether to output accurate Loss
        name: name.

    Returns:
        A tensor.
    """
    if context.executing_eagerly():
        raise RuntimeError("tf.k_means_centroids() is not compatible with "
                           "eager execution.")
    x = ops.convert_to_tensor(x, name="x")
    y = ops.convert_to_tensor(y, name="y")
    sum_square_y = ops.convert_to_tensor(sum_square_y, name="sum_square_y")
    if sum_square_x is not None:
        sum_square_x = ops.convert_to_tensor(sum_square_x, name="sum_square_x")
        use_actual_distance = True
    else:
        use_actual_distance = False

    if use_actual_distance:
        result = gen_npu_ops.k_means_centroids(x, y, sum_square_y, sum_square_x, use_actual_distance, name)
    else:
        result = gen_npu_ops.k_means_centroids_v2(x, y, sum_square_y, use_actual_distance, name)
    return result


def npu_onnx_graph_op(inputs, tout, model_path, name=None):
    """NPU implemented onnx graph operator"""
    output = gen_npu_ops.npu_onnx_graph_op(inputs, tout, model_path, name)
    return output

File: tf_adapter/ops/npu_ops.cc
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2019-2020. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "tensorflow/core/framework/common_shape_fns.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/framework/op.h"

namespace tensorflow {
using shape_inference::DimensionHandle;
using shape_inference::InferenceContext;
using shape_inference::ShapeHandle;
using shape_inference::UnchangedShape;

namespace {
REGISTER_OP("GeOp")
    .Input("inputs: Tin")
    .Attr("Tin: list(type) >= 0")
    .Output("outputs: Tout")
    .Attr("Tout: list(type) >= 0")
    .Attr("function: func")
    .Attr("data_format: { 'NHWC', 'NCHW', 'NDHWC', 'NCDHW', 'DHWCN', 'DHWNC'} = 'NHWC'")
    .SetIsStateful();

REGISTER_OP("DPOP")
    .Input("inputs: Tin")
    .Attr("Tin: list(type) >= 0")
    .Output("outputs: Tout")
    .Attr("Tout: list(type) >= 0")
    .Attr("function: func")
    .Attr("data_format: { 'NHWC', 'NCHW'} = 'NHWC'")
    .SetIsStateful();

REGISTER_OP("NPUInit").SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("LogTimeStamp").Attr("logid: string").Attr("notify: bool").SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("NPUShutdown").SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("LARS")
    .Input("inputs_w: T")
    .Input("inputs_g: T")
    .Input("weight_decay: float")
    .Output("outputs: T")
    .Attr("T: list(type) >= 1")
    .Attr("hyperpara: float = 0.001")
    .Attr("epsilon: float = 0.00001")
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      for (int i = 0; i < ((c->num_inputs() - 1) / 2); i++) {
        c->set_output(i, c->input(i));
      }
      return Status::OK();
    })
    .Doc(R"doc(
    Perform Lars on multi tensors. inputs_g have the same shape as `inputs_w`.

    Arguments
        inputs_w:    Tensors of weight.
        inputs_g:    Tensors of gradient.

    Output
        outputs:    Tensors with the same shape as `inputs_w`.
    )doc");

REGISTER_OP("LarsV2")
    .Input("input_weight: T")
    .Input("input_grad: T")
    .Input("weight_decay: T")
    .Input("learning_rate: T")
    .Output("output: T")
    .Attr("T: {float}")
    .Attr("hyperpara: float = 0.001")
    .Attr("epsilon: float = 0.00001")
    .Attr("use_clip: bool = false")
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(0));
      return Status::OK();
    })
    .Doc(R"doc(
    Perform LarsV2 on single output. input_weight have the same shape
    as `input_grad`.

    Arguments
        input_weight:    Tensor of weight.
        input_grad:      Tensor of gradient.
        weight_decay:    Tensor of weight_decay.
        learning_rate:   Tensor of learning_rate.
        use_clip:        Indicates whether to limit the coeff to acertain range.

    Output
        output:    Tensor with the same shape as `input_weight`.
    )doc");

Status OutfeedDequeueShapeFn(shape_inference::InferenceContext *c) {
  std::vector<PartialTensorShape> output_shapes;
  TF_RETURN_IF_ERROR(c->GetAttr("output_shapes", &output_shapes));
  if (static_cast<int>(output_shapes.size()) != c->num_outputs()) {
    return errors::InvalidArgument("`output_shapes` must be the same length as `output_types` (", output_shapes.size(),
                                   " vs. ", c->num_outputs());
  }
  for (size_t i = 0; i < output_shapes.size(); ++i) {
    shape_inference::ShapeHandle output_shape_handle;
    TF_RETURN_IF_ERROR(c->MakeShapeFromPartialTensorShape(output_shapes[i], &output_shape_handle));
    c->set_output(static_cast<int>(i), output_shape_handle);
  }
  return Status::OK();
}

REGISTER_OP("OutfeedEnqueueOp")
    .Input("inputs: Tin")
    .Attr("channel_name: string")
    .Attr("Tin: list(type) >= 0")
    .SetIsStateful()
    .SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("OutfeedDequeueOp")
    .Output("outputs: output_types")
    .Attr("channel_name: string")
    .Attr("output_types: list(type) >= 1")
    .Attr("output_shapes: list(shape) >= 1")
    .SetIsStateful()
    .SetShapeFn(OutfeedDequeueShapeFn);

REGISTER_OP("DropOutDoMask")
    .Input("x: T")
    .Input("mask: uint8")
    .Input("keep_prob: T")
    .Output("y: T")
    .Attr("T: {float16, float32}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(0));
      return Status::OK();
    });

REGISTER_OP("DropOutGenMask")
    .Input("shape: T")
    .Attr("T: {int64, int32}")
    .Input("prob: S")
    .Attr("S: {float, half}")
    .Output("output: uint8")
    .Attr("seed: int = 0")
    .Attr("seed2: int = 0")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      ShapeHandle unused;
      TF_RETURN_IF_ERROR(c->WithRankAtMost(c->input(1), 0, &unused));  // prob must be 0-d

      ShapeHandle inputShapeHandle;
      TF_RETURN_IF_ERROR(c->MakeShapeFromShapeTensor(0, &inputShapeHandle));

      int32 rank = InferenceContext::Rank(inputShapeHandle);
      if (rank == InferenceContext::kUnknownRank) {
        ShapeHandle out = c->UnknownShapeOfRank(1);
        c->set_output(0, out);
        return Status::OK();
      }

      bool unknownDimExist = false;
      for (int32 i = 0; i < rank; ++i) {
        DimensionHandle dimHandle = c->Dim(inputShapeHandle, i);
        int64 value = InferenceContext::Value(dimHandle);
        if (value == InferenceContext::kUnknownDim) {
          unknownDimExist = true;
          break;
        }
      }

      if (unknownDimExist) {
        ShapeHandle out = c->UnknownShapeOfRank(1);
        c->set_output(0, out);
        return Status::OK();
      }

      int64 bitCount = 0;
      if (rank != 0) {
        DimensionHandle inputDimHandle = c->NumElements(inputShapeHandle);
        bitCount = InferenceContext::Value(inputDimHandle);
      }

      // align to 128 and around up
      int64 n128Bits = bitCount / 128;
      if ((bitCount % 128) != 0) {
        n128Bits++;
      }

      // transfer 128 bit count to byte count if shape is full know
      int64 nBytes = n128Bits * 16;

      ShapeHandle out = c->Vector(nBytes);
      c->set_output(0, out);
      return Status::OK();
    });

REGISTER_OP("DropOutGenMaskV3")
    .Input("shape: T")
    .Attr("T: {int64, int32}")
    .Input("prob: S")
    .Attr("S: {float, half}")
    .Output("output: uint8")
    .Attr("seed: int = 0")
    .Attr("seed2: int = 0")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      ShapeHandle unused;
      // prob must be 0-d
      TF_RETURN_IF_ERROR(c->WithRankAtMost(c->input(1), 0, &unused));
      ShapeHandle input_shape_handle;
      TF_RETURN_IF_ERROR(c->MakeShapeFromShapeTensor(0, &input_shape_handle));
      if (!c->FullyDefined(input_shape_handle)) {
        ShapeHandle out = c->UnknownShape();
        c->set_output(0, out);
        return Status::OK();
      }
      int32 rank = InferenceContext::Rank(input_shape_handle);
      // [*batch, M, N] -> [*batch, N/16, M/16, 16, 16]
      if (rank >= 2) {
        DimensionHandle tmp_dim_handle = c->Dim(input_shape_handle, -1);
        int64 last_dim = InferenceContext::Value(tmp_dim_handle);
        tmp_dim_handle = c->Dim(input_shape_handle, -2);
        int64 second_last_dim = InferenceContext::Value(tmp_dim_handle);
        const int64 align = 16;
        if (last_dim % align == 0 && second_last_dim % align == 0) {
          last_dim /= align;
          second_last_dim /= align;
          tmp_dim_handle = c->MakeDim(last_dim);
          ShapeHandle out_shape_handle;
          TF_RETURN_IF_ERROR(c->ReplaceDim(input_shape_handle, -2, tmp_dim_handle, &out_shape_handle));
          tmp_dim_handle = c->MakeDim(second_last_dim);
          TF_RETURN_IF_ERROR(c->ReplaceDim(out_shape_handle, -1, tmp_dim_handle, &out_shape_handle));
          ShapeHandle tmp_shape_handle = c->Matrix(align, align);
          TF_RETURN_IF_ERROR(c->Concatenate(out_shape_handle, tmp_shape_handle, &out_shape_handle));
          c->set_output(0, out_shape_handle);
          return Status::OK();
        }
      }

      DimensionHandle input_dim_handle = c->NumElements(input_shape_handle);
      uint64 random_count = static_cast<uint64>(InferenceContext::Value(input_dim_handle));
      if (random_count > (INT64_MAX - 15)) {
        return errors::InvalidArgument("Required random count[", random_count, "] exceed INT64_MAX - 15");
      }
      // align to 16
      random_count = (random_count + 15) & (~15);
      ShapeHandle out = c->Vector(static_cast<int64>(random_count));
      c->set_output(0, out);
      return Status::OK();
    });

REGISTER_OP("BasicLSTMCell")
    .Input("x: T")
    .Input("h: T")
    .Input("c: T")
    .Input("w: T")
    .Input("b: T")
    .Output("ct: T")
    .Output("ht: T")
    .Output("it: T")
    .Output("jt: T")
    .Output("ft: T")
    .Output("ot: T")
    .Output("tanhct: T")
    .Attr("T: {float16, float32}")
    .Attr("keep_prob: float = 1.0")
    .Attr("forget_bias: float = 1.0")
    .Attr("state_is_tuple: bool = true")
    .Attr("activation: string = 'tanh'")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(2));
      c->set_output(1, c->input(1));
      c->set_output(2, c->input(2));
      c->set_output(3, c->input(2));
      c->set_output(4, c->input(2));
      c->set_output(5, c->input(2));
      c->set_output(6, c->input(2));
      return Status::OK();
    });

REGISTER_OP("BasicLSTMCellCStateGrad")
    .Input("c: T")
    .Input("dht: T")
    .Input("dct: T")
    .Input("it: T")
    .Input("jt: T")
    .Input("ft: T")
    .Input("ot: T")
    .Input("tanhct: T")
    .Output("dgate: T")
    .Output("dct_1: T")
    .Attr("T: {float16, float32}")
    .Attr("forget_bias: float = 1.0")
    .Attr("activation: string = 'tanh'")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      auto input_it_shape = c->input(4);
      auto hidden_size = c->Dim(input_it_shape, 1);
      auto batch_size = c->Dim(input_it_shape, 0);
      DimensionHandle output_size;
      TF_RETURN_IF_ERROR(c->Multiply(hidden_size, 4, &output_size));
      auto output_shape = c->MakeShape({batch_size, output_size});
      c->set_output(0, output_shape);
      c->set_output(1, c->input(2));
      return Status::OK();
    });

REGISTER_OP("BasicLSTMCellWeightGrad")
    .Input("x: T")
    .Input("h: T")
    .Input("dgate: T")
    .Output("dw: T")
    .Output("db: T")
    .Attr("T: {float16, float32}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      auto input_x_shape = c->input(0);
      auto input_h_shape = c->input(1);
      auto input_dgate_shape = c->input(2);
      auto four_hidden_size = c->Dim(input_dgate_shape, 1);
      auto hidden_size = c->Dim(input_h_shape, 1);
      auto input_size = c->Dim(input_x_shape, 1);
      DimensionHandle output_size;
      TF_RETURN_IF_ERROR(c->Add(hidden_size, input_size, &output_size));
      auto output_dw_shape = c->MakeShape({output_size, four_hidden_size});
      auto output_db_shape = c->MakeShape({four_hidden_size});
      c->set_output(0, output_dw_shape);
      c->set_output(1, output_db_shape);
      return Status::OK();
    });

REGISTER_OP("BasicLSTMCellInputGrad")
    .Input("dgate: T")
    .Input("w: T")
    .Output("dxt: T")
    .Output("dht: T")
    .Attr("T: {float16, float32}")
    .Attr("keep_prob: float = 1.0")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      auto input_dgate_shape = c->input(0);
      auto input_w_shape = c->input(1);
      auto four_hidden_size = c->Dim(input_dgate_shape, 1);
      auto batch_size = c->Dim(input_dgate_shape, 0);
      auto input_hidden_size = c->Dim(input_w_shape, 0);
      DimensionHandle output_hidden_size;
      TF_RETURN_IF_ERROR(c->Divide(four_hidden_size, 4, true, &output_hidden_size));
      auto output_dht_shape = c->MakeShape({batch_size, output_hidden_size});
      DimensionHandle output_input_size;
      TF_RETURN_IF_ERROR(c->Subtract(input_hidden_size, output_hidden_size, &output_input_size));
      auto output_dxt_shape = c->MakeShape({batch_size, output_input_size});
      c->set_output(0, output_dxt_shape);
      c->set_output(1, output_dht_shape);
      return Status::OK();
    });

REGISTER_OP("AdamApplyOneAssign")
    .Input("input0: T")
    .Input("input1: T")
    .Input("input2: T")
    .Input("input3: T")
    .Input("input4: T")
    .Input("mul0_x: T")
    .Input("mul1_x: T")
    .Input("mul2_x: T")
    .Input("mul3_x: T")
    .Input("add2_y: T")
    .Attr("T: {float16, float32}")
    .SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("LambApplyOptimizerAssign")
    .Input("input0: T")
    .Input("input1: T")
    .Input("input2: T")
    .Input("input3: T")
    .Input("mul0_x: T")
    .Input("mul1_x: T")
    .Input("mul2_x: T")
    .Input("mul3_x: T")
    .Input("add2_y: T")
    .Input("steps: T")
    .Input("do_use_weight: T")
    .Input("weight_decay_rate: T")
    .Output("update: T")
    .Output("output1: T")
    .Output("output2: T")
    .Attr("T: {float16, float32}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(0));
      c->set_output(1, c->input(1));
      c->set_output(2, c->input(2));
      return Status::OK();
    });

REGISTER_OP("LambApplyWeightAssign")
    .Input("input0: T")
    .Input("input1: T")
    .Input("input2: T")
    .Input("input3: T")
    .Input("input4: T")
    .Output("output0: T")
    .Attr("T: {float16, float32}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(4));
      return Status::OK();
    });

REGISTER_OP("AdamApplyOneWithDecayAssign")
    .Input("input0: T")
    .Input("input1: T")
    .Input("input2: T")
    .Input("input3: T")
    .Input("input4: T")
    .Input("mul0_x: T")
    .Input("mul1_x: T")
    .Input("mul2_x: T")
    .Input("mul3_x: T")
    .Input("mul4_x: T")
    .Input("add2_y: T")
    .Attr("T: {float16, float32}")
    .SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("NpuOnnxGraphOp")
    .Input("inputs: Tin")
    .Attr("Tin: list(type) >= 0")
    .Output("outputs: Tout")
    .Attr("Tout: list(type) >= 0")
    .Attr("model_path: string")
    .SetShapeFn(shape_inference::UnknownShape);

REGISTER_OP("KMeansCentroids")
    .Input("x: T")
    .Input("y: T")
    .Input("sum_square_y: T")
    .Input("sum_square_x: T")
    .Output("segment_sum: T")
    .Output("segment_count: T")
    .Output("kmean_total_sum: T")
    .Attr("T: {float32}")
    .Attr("use_actual_distance: bool = false")
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      auto input_y_shape = c->input(1);
      auto n = c->Dim(input_y_shape, 0);
      auto d = c->Dim(input_y_shape, 1);
      c->set_output(0, c->MakeShape({n, d}));
      c->set_output(1, c->MakeShape({n, 1}));
      c->set_output(2, c->MakeShape({1}));
      return Status::OK();
    });

REGISTER_OP("KMeansCentroidsV2")
    .Input("x: T")
    .Input("y: T")
    .Input("sum_square_y: T")
    .Output("segment_sum: T")
    .Output("segment_count: T")
    .Output("kmean_total_sum: T")
    .Attr("T: {float32}")
    .Attr("use_actual_distance: bool = false")
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      auto input_y_shape = c->input(1);
      auto n = c->Dim(input_y_shape, 0);
      auto d = c->Dim(input_y_shape, 1);
      c->set_output(0, c->MakeShape({n, d}));
      c->set_output(1, c->MakeShape({n, 1}));
      c->set_output(2, c->MakeShape({1}));
      return Status::OK();
    });


REGISTER_OP("FileConstant")
    .Output("y: dtype")
    .Attr("file_path: string = ''")
    .Attr("file_id: string = ''")
    .Attr("shape: list(int)")
    .Attr("dtype: {float32, float16, int8, int16, uint16, uint8, int32, int64, uint32, uint64, bool, double}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      std::vector<int32_t> output_shape;
      TF_RETURN_IF_ERROR(c->GetAttr("shape", &output_shape));
      size_t rank = output_shape.size();
      std::vector<DimensionHandle> out_dims(rank);
      for (size_t i = 0UL; i < rank; i++) {
        out_dims[i] = c->MakeDim(shape_inference::DimensionOrConstant(output_shape[i]));
      }
      c->set_output(0, c->MakeShape(out_dims));
      return Status::OK();
    });
}  // namespace
}  // namespace tensorflow
